truffle/src/main/java/org/truffleruby/core/queue/UnsizedQueue.java

```diff
@@ -0,0 +1,178 @@
+    private Item takeEnd;
+    private int size;
+
+    public void add(Object item) {
+        lock.lock();
+
+        try {
+            final Item newItem = new Item(item);
+            if (addEnd != null) {
+                addEnd.setNextToTake(newItem);
+            }
+            addEnd = newItem;
+            if (takeEnd == null) {
+                takeEnd = addEnd;
+            }
+            size++;
+            canTake.signal();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object poll() {
+        lock.lock();
+
+        try {
+            if (takeEnd == null) {
+                return null;
+            } else {
+                return doTake();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object take() throws InterruptedException {
+        lock.lock();
+
+        try {
+            while (takeEnd == null) {
+                canTake.await();
+            }
+
+            return doTake();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Object poll(long timeout, TimeUnit unit) throws InterruptedException {
+        lock.lock();
+
+        try {
+            while (takeEnd == null) {
+                // We'll repeat the same timeout period again if we skip in this loop, but it's not a guarantee of anyway
+                canTake.await(timeout, unit);
+            }
+
+            return doTake();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private Object doTake() {
+        assert lock.isHeldByCurrentThread();
+        final Object item = takeEnd.getItem();
+        takeEnd = takeEnd.getNextToTake();
+        if (takeEnd == null) {
+            addEnd = null;
+        }
+        size--;
+        return item;
+    }
+
+    public int size() {
+        lock.lock();
+
+        try {
+            return size;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    public void clear() {
+        lock.lock();
+
+        try {
+            addEnd = null;
+            takeEnd = null;
+            size = 0;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public int getNumberWaitingToTake() {
+        lock.lock();
+
+        try {
+            return lock.getWaitQueueLength(canTake);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Collection<Object> getContents() {
+        lock.lock();
+
+        final Collection<Object> objects = new ArrayList<>();
+
+        try {
+            Item iterator = takeEnd;
+
+            while (iterator != null) {
+                objects.add(iterator.getItem());
+                iterator = iterator.getNextToTake();
+            }
+        } finally {
+            lock.unlock();
+        }
+
+        return objects;
+    }
+
+    private class Item {
+
+        private final Object item;
+        private Item nextToTake;
```

Comment:
They are taken following the `next` pointers. A small picture like this would clarify simply:
```
head=v1 -> v2 -> ... -> tail=vn
pop from head, push to tail
```